building/build-acis/flannel-monitor/src/flannel-monitor-collector/main/flannel-monitor-collector.go

```diff
@@ -0,0 +1,206 @@
+		return
+	}
+	podlist, err := clientset.CoreV1().Pods(namespace).List(v1.ListOptions{LabelSelector: "app=flannel-monitor"})
+	if err != nil {
+		scrapeCheck.Reset()
+		additional_metrics = nil
+		collectCheck.Set(0)
+		dupCheck.Set(0)
+		log.Printf("could not fetch list of flannel-monitor apps")
+		return
+	}
+
+	check_map := map[string]prometheus.Gauge{}
+	not_found := map[string]struct{}{}
+	for _, item := range nodelist.Items {
+		nodeIP := ""
+		for _, addr := range item.Status.Addresses {
+			if addr.Type == v12.NodeInternalIP {
+				if nodeIP != "" && nodeIP != addr.Address {
+					scrapeCheck.Reset()
+					additional_metrics = nil
+					collectCheck.Set(0)
+					dupCheck.Set(0)
+					log.Printf("got duplicate IPs for node: %s", item.Name)
+					return
+				}
+				nodeIP = addr.Address
+			}
+		}
+		if nodeIP == "" {
+			scrapeCheck.Reset()
+			additional_metrics = nil
+			collectCheck.Set(0)
+			dupCheck.Set(0)
+			log.Printf("could not get IP for node: %s", item.Name)
+			return
+		}
+		check_map[nodeIP] = scrapeCheck.With(prometheus.Labels{
+			"monitor_hostip": nodeIP,
+		})
+		not_found[nodeIP] = struct{}{}
+	}
+
+	dupPass := float64(1)
+
+	var result []*io_prometheus_client.MetricFamily
+	textParser := &expfmt.TextParser{}
+
+	for _, item := range podlist.Items {
+		if item.Status.Phase != "Running" {
+			continue
+		}
+		hostIP := item.Status.HostIP
+		podIP := item.Status.PodIP
+		if _, present := not_found[hostIP]; !present {
+			log.Printf("unexpected or duplicate entry: %s", hostIP)
+			dupPass = 0
+		}
+		delete(not_found, hostIP)
+
+		resp, err := client.Get(fmt.Sprintf("http://%s/metrics", podIP))
+		pass := float64(0)
+		if err != nil {
+			log.Printf("failed to fetch metrics from %s: %v", podIP, err)
+		} else {
+			metrics, err := textParser.TextToMetricFamilies(resp.Body)
+			if err != nil {
+				log.Printf("failed to parse metrics from to %s: %v", podIP, err)
+			} else {
+				for _, family := range metrics {
+					result = append(result, family)
+				}
+				pass = 1
+			}
+		}
+		check_map[hostIP].Set(pass)
+	}
+
+	for k, _ := range not_found {
+		check_map[k].Set(0)
+	}
+
+	dupCheck.Set(dupPass)
+	collectCheck.Set(1)
+
+	additional_metrics = result
+}
+
+func gather() ([]*io_prometheus_client.MetricFamily, error) {
+	adl := additional_metrics
+	gathered, err := registry.Gather()
+	if err != nil {
+		return nil, err
+	}
+	temp := make([]*io_prometheus_client.MetricFamily, 0, len(adl) + len(gathered))
+	temp = append(temp, gathered...)
+	temp = append(temp, adl...)
+	return temp, nil
+}
+
+func loop(clientset *kubernetes.Clientset, namespace string, client *http.Client, stopChannel <-chan struct{}) {
+	for {
+		next_cycle_at := time.Now().Add(time.Second * 15)
+		cycle(clientset, namespace, client)
+
+		delta := next_cycle_at.Sub(time.Now())
+		if delta < time.Second {
+			delta = time.Second
+		}
+
+		select {
+		case <-stopChannel:
+			break
+		case <-time.After(delta):
+		}
```

Comment:
Is this equivalent to using a [Ticker](https://golang.org/pkg/time/#Ticker)?

Reply:
I don't think so. If I'm reading the code right, an instance of Ticker will drop missed ticks, which would mean that a 16 second cycle would be rounded up to a 30 second cycle, instead of a 17 second cycle. This is supposed to just be a rate-limit, so this kind of rounding is unhelpful.